include/device.h

```diff
@@ -305,6 +341,84 @@ struct device {
 #endif
 };
 
+/**
+ * @brief Get the handle for a given device
+ *
+ * @param dev the device for which a handle is desired.
+ *
+ * @return the handle for the device, or DEVICE_HANDLE_NULL if the
+ * device does not have an associated handle.
+ */
+static inline device_handle_t
+device_handle_get(const struct device *dev)
+{
+	device_handle_t ret = DEVICE_HANDLE_NULL;
+	extern const struct device __device_start[];
+
+	/* TODO: If/when devices can be constructed that are not part of the
+	 * fixed sequence we'll need another solution.
+	 */
+	if (dev != NULL) {
+		ret = 1 + (device_handle_t)(dev - __device_start);
+	}
+
+	return ret;
+}
+
+/**
+ * @brief Get the device corresponding to a handle.
+ *
+ * @param dev_handle the device handle
+ *
+ * @return the device that has that handle, or a null pointer if @p
+ * dev_handle does not identify a device.
+ */
+static inline const struct device *
+device_from_handle(device_handle_t dev_handle)
+{
+	extern const struct device __device_start[];
+	extern const struct device __device_end[];
+	const struct device *dev = NULL;
+	size_t numdev = __device_end - __device_start;
+
+	if ((dev_handle > 0) && ((size_t)dev_handle < numdev)) {
+		dev = &__device_start[dev_handle - 1];
+	}
+
+	return dev;
+}
+
+/**
+ * @brief Get the set of devices on which a given device depends
+ *
+ * @param dev the device for which dependencies are desired.
+ *
+ * @param count pointer to a place to store the number of devices provided at
+ * the returned pointer.  The value is not set if the call returns a null
+ * pointer.  The value may be set to zero.
+ *
+ * @return a pointer to a sequence of @p *count device handles, or a null
+ * pointer if @p dh does not provide dependency information.
+ */
+static inline const device_handle_t *
+device_get_requires_handles(const struct device *dev,
```

Comment:
how about `device_dependency_get(..)`

Reply:
The problem is that when we support dynamic dependencies this API won't work, because we won't have all devices in a contiguous array.  We'll need a callback-based API to iterate over both static and dynamic dependencies.

But I can make this `device_requires_handles_get()` which would handle the static fetch with naming closer to the proposed standard.

Reply:
@nashif Changed to `device_required_handles_get()`.  Turns out when this was reverted the test code was inadvertently left in tree because it had been patched subsequent to the original merge.

Reply:
the description of the API is `Get the set of devices on which a given device depends`, so `dependency_get` would have reflected that. 

I see you changed that to `device_required_handles_get`, this is fine I guess (better than device_requires_handles_get).

Reply:
This had to be rebased to resolve a massive conflict with #31880 so I changed the description to reflect that it's limited to devicetree required dependencies.  With #31880 in I can now add API to record dynamic dependencies and add a solution for "`device_dependency_get()`"